src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
+        // If a tx is to-be-replaced, remove any of its descendants so they can't fee-bump anything.
+        // this case should be rare as the wallet won't normally attempt to replace transactions
+        // with descendants.
+        const bool remove = to_be_replaced.find(txid) != to_be_replaced.end();
+        CTxMemPool::setEntries descendants;
+        mempool.CalculateDescendants(txiter, descendants);
+
+        for (const auto& desc_txiter : descendants) {
+            auto desc_it{entries_by_txid.find(desc_txiter->GetTx().GetHash())};
+            if (desc_it != entries_by_txid.end()) {
+                if (remove) {
+                    entries_by_txid.erase(desc_it);
+                } else {
+                    cached_descendants.push_back(desc_it);
+                }
+            }
+        }
+        if (!remove) descendant_set_by_txid.emplace(std::make_pair(txid, cached_descendants));
+    }
+    // Release the mempool lock; we now have all the information we need for a subset of the entries
+    // we care about. We will solely operate on the MockMempoolEntry map from now on.
+    assert(entries.size() == entries_by_txid.size());
+    assert(entries.size() == descendant_set_by_txid.size());
+    assert(in_block.empty());
+}
+
+// Compare by ancestor feerate, then iterator
+struct AncestorFeerateComparator
+{
+    template<typename I>
+    bool operator()(const I& a, const I& b) const {
+        const CFeeRate a_feerate(a->second.GetModFeesWithAncestors(), a->second.GetSizeWithAncestors());
+        const CFeeRate b_feerate(b->second.GetModFeesWithAncestors(), b->second.GetSizeWithAncestors());
+        if (a_feerate != b_feerate) {
+            return a_feerate > b_feerate;
+        }
+        return &(*a) > &(*b);
+    }
+};
+
+void MiniMiner::DeleteAncestorPackage(const std::set<MockEntryMap::iterator, IteratorComparator>& ancestors)
+{
+    for (const auto& anc : ancestors) {
+        auto vec_it = std::find(entries.begin(), entries.end(), anc);
+        assert(vec_it != entries.end());
+        entries.erase(vec_it);
+        entries_by_txid.erase(anc);
+    }
+}
+
+void MiniMiner::BuildMockTemplate(const CFeeRate& target_feerate)
+{
+    while (!entries_by_txid.empty()) {
+        // Sort again, since transaction removal may change some entries' ancestor feerates.
+        std::sort(entries.begin(), entries.end(), AncestorFeerateComparator());
+
+        // Pick highest ancestor feerate entry.
+        auto best_iter = entries.begin();
+        assert(best_iter != entries.end());
+        const auto ancestor_package_size = (*best_iter)->second.GetSizeWithAncestors();
+        const auto ancestor_package_fee = (*best_iter)->second.GetModFeesWithAncestors();
+        // Stop here. Everything that didn't "make it into the block" has bumpfee.
+        if (best_iter == entries.end() || ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
```

Comment:
Simplification, we just asserted that it's not `entries.end()`
```suggestion
        if (ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
```